src/libraries/System.Text.Json/source_generation/src/JsonSourceGenerator.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace System.Text.Json.SourceGeneration
+{
+    [Generator]
+    public sealed class JsonSourceGenerator : ISourceGenerator
+    {
+        public Dictionary<string, Type>? SerializableTypes { get; private set; }
+
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new JsonSerializableSyntaxReceiver());
+        }
+
+        public void Execute(GeneratorExecutionContext executionContext)
```

Comment:
I kind of like the pattern established by

* [https://github.com/dotnet/runtime/tree/main/src/libraries/System.Private.CoreLib/generators](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Private.CoreLib/generators)

and also used in [https://github.com/dotnet/runtime/pull/51064](https://github.com/dotnet/runtime/pull/51064)

Split the implementation into 2 classes:

* .Parser
* .Emitter

And then you can easily see the separate of concerns - one interprets the code, and builds up the information necessary to generate code. The Emitter then takes that information and turns it into a C# string.

Reply:
I think it's a good pattern. The logic in this generator is somewhat similar, just in one class and split in two files. Logic in `JsonSourceGeneratorHelper` builds up the metadata we need to emit the source, while `JsonSourceGeneratorHelper.Generate`  emits the source.

I'd like to align more closely with the .Parser/.Emitter pattern as you described it, but preferably in a follow up PR at this point. I've taken a note to do this here - https://github.com/dotnet/runtimelab/projects/1#column-11545555.

--- 

> The Emitter then takes that information and turns it into a C# string.

Is the goal to produce a single string/file; or could the emitter emit multiple? Multiple (one string for each type) is preferred in the case of JSON, so that one can easily debug metadata for a specific type.

Reply:
> Is the goal to produce a single string/file

I think the Emitter can produce as many files as it needs to.